File: pkg-py/tests/test_indeterminate.py
"""on_error='indeterminate': one unreachable id does not sink the batch."""

import pytest

import biogate
import biogate._remote as remote


def test_default_raises_on_a_remote_failure(monkeypatch, tmp_path):
    monkeypatch.setenv("BIOGATE_CACHE_DIR", str(tmp_path))

    def _boom(url, timeout=30):
        raise remote.RemoteError("connection refused")

    monkeypatch.setattr(remote, "_http_get", _boom)
    with pytest.raises(remote.RemoteError):
        biogate.check_id("MONDO:0005148", source_db="mondo", how="remote")


def test_indeterminate_isolates_the_failure(monkeypatch, tmp_path):
    monkeypatch.setenv("BIOGATE_CACHE_DIR", str(tmp_path))

    # One id resolves, the other always fails: the good one is still checked.
    def _mixed(url, timeout=30):
        if "obo_id=MONDO:0005148" in url:
            return 200, {"page": {"totalElements": 1}}
        raise remote.RemoteError("connection refused")

    monkeypatch.setattr(remote, "_http_get", _mixed)
    res = biogate.check_id(
        ["MONDO:0005148", "MONDO:0000001"],
        source_db="mondo",
        how="remote",
        on_error="indeterminate",
    )
    assert res[0].valid is True
    assert res[0].error is None
    # The unreachable id is indeterminate: valid None with a reason, not False.
    assert res[1].valid is None
    assert res[1].error is not None
    assert "connection refused" in res[1].error


def test_indeterminate_does_not_cache_the_failure(monkeypatch, tmp_path):
    monkeypatch.setenv("BIOGATE_CACHE_DIR", str(tmp_path))
    calls = {"n": 0}

    def _boom(url, timeout=30):
        calls["n"] += 1
        raise remote.RemoteError("connection refused")

    monkeypatch.setattr(remote, "_http_get", _boom)
    biogate.check_id("MONDO:0005148", "mondo", how="remote", on_error="indeterminate")
    biogate.check_id("MONDO:0005148", "mondo", how="remote", on_error="indeterminate")
    # A second run refetches, so the failure was never persisted as a verdict.
    assert calls["n"] == 2


def test_malformed_input_is_invalid_not_indeterminate(monkeypatch, tmp_path):
    monkeypatch.setenv("BIOGATE_CACHE_DIR", str(tmp_path))

    def _boom(url, timeout=30):
        raise remote.RemoteError("connection refused")

    monkeypatch.setattr(remote, "_http_get", _boom)
    # "mondo:5148" is malformed; it is invalid regardless of the network, and its
    # suggestion just cannot be confirmed, so it is not indeterminate.
    res = biogate.check_id(
        "mondo:5148", source_db="mondo", how="remote", on_error="indeterminate"
    )[0]
    assert res.valid is False
    assert res.error is None


def test_invalid_on_error_rejected():
    with pytest.raises(ValueError, match="on_error"):
        biogate.check_id("MONDO:0005148", "mondo", on_error="nonsense")

File: pkg-r/tests/testthat/test-indeterminate.R
# on_error = "indeterminate": one unreachable id does not sink the batch. The
# test transport aborts with class biobouncer_error_remote to stand in for a network
# failure that has exhausted its retries.

.fail_transport <- function(url, timeout) {
  cli::cli_abort("connection refused", class = "biobouncer_error_remote")
}

.mixed_transport <- function(url, timeout) {
  if (grepl("obo_id=MONDO:0005148", url, fixed = TRUE)) {
    list(status = 200, body = list(page = list(totalElements = 1)))
  } else {
    cli::cli_abort("connection refused", class = "biobouncer_error_remote")
  }
}

test_that("the default raises on a remote failure", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .fail_transport)
  expect_error(
    check_id("MONDO:0005148", source_db = "mondo", how = "remote"),
    class = "biobouncer_error_remote"
  )
})

test_that("indeterminate isolates the failure to one id", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .mixed_transport)
  res <- check_id(
    c("MONDO:0005148", "MONDO:0000001"),
    source_db = "mondo",
    how = "remote",
    on_error = "indeterminate"
  )
  expect_true(res$valid[1])
  expect_true(is.na(res$error[1]))
  # the unreachable id is indeterminate: NA verdict with a reason, not FALSE.
  expect_true(is.na(res$valid[2]))
  expect_match(res$error[2], "connection refused")
})

test_that("a malformed input stays invalid, not indeterminate", {
  withr::local_envvar(BIOBOUNCER_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biobouncer.remote_transport = .fail_transport)
  res <- check_id(
    "mondo:5148",
    source_db = "mondo",
    how = "remote",
    on_error = "indeterminate"
  )
  expect_false(res$valid)
  expect_true(is.na(res$error))
})

test_that("an invalid on_error value is rejected", {
  expect_error(
    check_id("MONDO:0005148", source_db = "mondo", on_error = "nonsense"),
    class = "biobouncer_error_invalid_on_error"
  )
})
